### PR TITLE
Update 14-writing-functions.md

### PR DESCRIPTION
**Pull request by lizcw**
Replaced last challenge (Encapsulating Data Analysis) with one which builds on previous lessons and provides a demonstrably useful method.
(NB, this is for completion of instructor training)

### CODE REVIEW

File: _episodes/14-writing-functions.md [hunk 1]
```diff
@@ -77,17 +77,14 @@ print_date(1871, 3, 19)
 ~~~
 {: .output}
 
-*   Via [Twitter](https://twitter.com/minisciencegirl/status/693486088963272705):
```
Comment (ntmoore) [inline _episodes/14-writing-functions.md]: I like this analogy and would rather is not be removed from the lesson.  Is there a specific reason you think it should be cut?

Comment (lizcw) [inline _episodes/14-writing-functions.md]: It doesn't really make sense to me - I actually thought it had been left in by error as the twitter discussion doesn't really reflect the statement it's linked to and is hard to follow.  I don't think calling arguments "ingredients" helps the understanding, in fact, ingredients imply specificity whereas arguments allow for flexibility.   There are references in an earlier lesson to the correct nomenclature (02-variables) so a deviation at this point could create confusion.

Comment (ntmoore) [inline _episodes/14-writing-functions.md]: If I am making a quiche, the ingredients for the filling are certainly eggs and milk, but also some sort of green vegetable, some sort of cheese, and perhaps some sort of meat.  While the eggs and milk are strictly specified, there is flexibility in other ingredients.  While the analogy of recipes to function definitions is imperfect, I think many people learn by extening known patterns to new concepts, (eg, "Oh, this [new idea] is like [another old idea] that I already understand.  Shayer and Adey call this "bridging" or "transfer" - [see this pdf](http://modeling.asu.edu/modeling/AdeyP-CognitiveAccel2011.pdf) )

Again, I would prefer to keep this as an anecdote at least.  I've noticed the writing in this section is quite terse.  Perhaps if it was presented explicity as an anecdote with explaination...

File: _episodes/14-writing-functions.md [hunk 2]
```diff
@@ -264,82 +262,77 @@ result of call is: None
...
-> 3. How would you generalize this function
->    if you did not know beforehand which specific years occurred as columns in the data?
->    For instance, what if we also had data from years ending in 1 and 9 for each decade?
->    (Hint: use the columns to filter out the ones that correspond to the decade,
->    instead of enumerating them in the code.) 
->
 > > ## Solution
 > >
-> > 1. 
+> > 1. This requires knowledge of the column headings so is of limited re-use
```
Comment (ntmoore) [inline _episodes/14-writing-functions.md]: I think there are a few problems that are scattered through the lessons that use column headings.  I thinnk we should retain this variation. (Not all of the exercises are required of course.)

**Comment by k8hertweck** [comment]
Thanks so much to for the PR, @lizcw! 

This is a tricky PR to discuss because it includes quite a few changes throughout the length of the module (with the main change described as replacing an exercise). How do you feel about the rest of the changes, @ntmoore? If the rest of it looks OK, here's what I propose as a course of action: Modify the PR to keep the analogy and merge. Open a separate PR specifically about the analogy and solicit some other opinions (perhaps from maintainers of other Python lessons?), so you can decide whether to keep it or develop the language a bit more.

Let me know if there's anything I can do to help!

**Comment by ntmoore** [comment]
I havn't worked through the exercise, but was open to accepting the pr aside from  the deletion of the cooking example.  Final exams right now.  If you want to modify the pr to be two separate pieces, one including the exercise and another starting a discussion of the cooking example, that's fine with me (I'll merge the coding example...)  

I do appreciate your contributions @lizcw!  Community contributions are what make these materials.

**Comment by lizcw** [comment]
As advised, reverting changes. Will open a new branch for replacement challenge example.
